### ISSUE TITLE
datastack: loading data that's not on a stack file

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#727]
I'm using the datastack module for the first time and I'm finding that it does not work as I thought it would. That is fine - it does what it should according to the docs. Still, I think more flexibility to load data would be useful. In that sense, this is a feature request, not a bug report.
However, I'm listing several aspects here, because they are all related, and some are current bug, in the sense that the error message is not good.

Feature request:
I can load multiple files with the stack syntax (`@filename.lis`), however working in python I don't see why I should write an external file. I would like to pass in a list of filenames.

Feature request:
Assign ids. Is there a way to assign an id to a dataset that I add to a datastack, e.g. call it with a string? The code indicates that such a possibility exists (https://github.com/sherpa/sherpa/blob/c1a48243bc8fe53123098644710bbf3c2f7af6a8/sherpa/astro/datastack/utils.py#L164) but I can't figure out to make it work. Even after looking at that piece of code for over an hour, I still don't understand how `id` can ever by anything else than `None` and what lines L164-L169 in that file are good for.

I naively attempted to just pass in a string, but that does not work:
```
ds2 = dsmod.DataStack()
ds2.load_data('idstring', filename)

UnboundLocalError                         Traceback (most recent call last)
<ipython-input-93-bfe740cadfec> in <module>
      1 ds2 = dsmod.DataStack()
----> 2 ds2.load_data('idstring', filename)

/nfs/melkor/d1/guenther/soft/anaconda/envs/ciao412/lib/python3.7/site-packages/sherpa/astro/datastack/utils.py in _load(self, *args, **kwargs)
    162             args = args[1:]
    163 
--> 164         if id is not None:
    165             if self._default_instance:
    166                 self._load_func(load_func, id, arg, *args, **kwargs)

UnboundLocalError: local variable 'id' referenced before assignment
```

So, I figures, I can pass in an id as an argument, but that does not work either:
```
ds2 = dsmod.DataStack()
ds2.load_data(pifiles[0], id='idstring')

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/nfs/melkor/d1/guenther/soft/anaconda/envs/ciao412/lib/python3.7/site-packages/sherpa/astro/datastack/utils.py in _load(self, *args, **kwargs)
    175             for file in files:
--> 176                 self._load_func(load_func, file, *args, **kwargs)
    177         except:

/nfs/melkor/d1/guenther/soft/anaconda/envs/ciao412/lib/python3.7/site-packages/sherpa/astro/datastack/ds.py in _load_func(self, func, *args, **kwargs)
    538         logger.info('Loading dataset id {0}'.format(dataid))
--> 539         func(dataid, *args, **kwargs)
    540 

TypeError: load_data() got multiple values for argument 'id'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-94-f06c1a6df39a> in <module>
      1 ds2 = dsmod.DataStack()
----> 2 ds2.load_data(pifiles[0], id='idstring')

/nfs/melkor/d1/guenther/soft/anaconda/envs/ciao412/lib/python3.7/site-packages/sherpa/astro/datastack/utils.py in _load(self, *args, **kwargs)
    176                 self._load_func(load_func, file, *args, **kwargs)
    177         except:
--> 178             self._load_func(load_func, arg, *args, **kwargs)
    179 
    180     _load.__name__ = load_func.__name__

/nfs/melkor/d1/guenther/soft/anaconda/envs/ciao412/lib/python3.7/site-packages/sherpa/astro/datastack/ds.py in _load_func(self, func, *args, **kwargs)
    537 
    538         logger.info('Loading dataset id {0}'.format(dataid))
--> 539         func(dataid, *args, **kwargs)
    540 
    541         self._add_dataset(dataid)

TypeError: load_data() got multiple values for argument 'id'
```

That makes me wonder what the case `len(args) > 1` can ever work? 

I've spend an afternoon digging, but unfortunately it's hard to follow the code and understand all cases in a wrapper to a function that monkey-patches sherpa's load functions, which already have a variable number of arguments.
Thus, I'm leaving this here as an issue so that my digging is at least recorded somewhere.

cc: @taldcroft

**Comment by olaurino** [comment]
Looking at the test cases: https://github.com/sherpa/sherpa/blob/master/sherpa/astro/datastack/tests/test_datastack.py

I don't see any tests for either requirement, i.e. setting an ID for data sets and listing files rather than writing to (and then reading from) a stack file.

While the code is from almost a decade ago and I'd question some of my choices (why in the world did I use `'/'.join` rather than `os.path.join`?!) I remember that I worked on this in a best case scenario: I could collect the requirements, get people to sign off on the collected requirements, on the design, and even on the test plan, and as an added bonus I was working with existing code from @taldcroft. The tests show this and are pretty comprehensive, and possibly easier to follow than the code.

In our internal documentation (this is pre-git, pre-confluence, pre-dokuwiki) I can't find a reference to either requirements. It might be that @taldcroft's original code supported setting an ID for datasets, but that my adaptation, lacking the requirement, introduced a regression.

**Comment by olaurino** [comment]
Wait, I was rereading your examples... isn't this your case?
https://github.com/sherpa/sherpa/blob/master/sherpa/astro/datastack/tests/test_datastack.py#L115

I thought you were trying to set the ID through a pattern as the datastack API would allow you to do with models, but now I see you are actually setting them individually, and that should be allowed.

**Comment by hamogu** [comment]
No, that's subtly different. I'm trying to use it on an *instance* of a datatstack, while in L115 `datastack` is the module and "myid" is the name of the datastack instance this file is read into, not the id of the dataset. I'm trying to read into a particular datastack instance (identified by the instance) **and** assign a name to the dataset.

**Comment by olaurino** [comment]
That makes sense. I am not a big fan of the fact that there are two ways of doing the same thing, but that's one of the most basic sherpa requirements. Today I would probably heavily parameterize the tests to try and catch as many edge cases as possible.

**Comment by hamogu** [comment]
But thank you for providing the background to me. I don't know if @taldcroft 's original implementation allowed for that. I think the datastack module is not used as much as it should, and using it in an OO way is clearly not the original intention. I'll keep using to for the science project I'm working on right now and I'll report back here on what I think after getting used to the interface more and, if needed, I'll open a PR.

**Comment by olaurino** [comment]
Actually, the OO interface was part of the design, I think I just missed a few cases. I'll see with @wmclaugh if it makes sense to import the original internal documentation (text files) in the applications we now use. That might help with extending/fixing the current implementation.

**Comment by hamogu** [comment]
I just discovered `_add_dataset(id)`. I don't see why that could not be a public method. I can load my data in any way I like and then just add it to the datastack.

In practice, we should not just rename `_add_dataset`  to remove the leading underscore, but probably add a public method like this:
```
def add_dataset(id):
    if id in ui.list_data_ids():
        self._add_dataset(id)
    else:
        raise UsefullExceptionHere...
```

**Comment by hamogu** [comment]
A `load_pha2` method might also be useful.

(I'll keep collecting ideas here as I'm using datastack in a real science project with grating data.)
